### PR TITLE
Update README and Dockerfile

### PR DESCRIPTION
**Pull request by adamreeve**
This PR includes a few tidy ups and fixes to examples in the README:
* Update versions of Arrow and DataFusion referenced in the README
* Update the version of datafusion-java used
* Update the Dockerfile so that jshell works again and update the section about using jshell in the README
* Fix invalid references to `ExecutionContexts` that should be `SessionContexts`
* Update the example code to include the full class definition and required import statements to make it easier to use

Some of this work is thanks to @pavlovic-ivan and was originally contributed in the G-Research fork

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM openjdk:11-jdk-slim-bullseye
+FROM openjdk:21-jdk-slim-bullseye
```
Comment (Jimexist) [inline Dockerfile:1]: maybe use alternative distro other than openjdk?

Comment (adamreeve) [reply Dockerfile:1]: What's the issue with OpenJDK, and do you have a recommendation? I don't usually do a lot of Java development to be honest so I'm not very familiar with the alternatives :sweat_smile:

Comment (Jimexist) [reply Dockerfile:1]: it's not about the jdk version but more of an issue with dockerfile and docker image

i'd personally use https://hub.docker.com/_/amazoncorretto

Comment (Jimexist) [reply Dockerfile:1]: if you look at https://hub.docker.com/_/openjdk it shows it's deprecated @adamreeve

Comment (adamreeve) [reply Dockerfile]: :+1: makes sense thanks, I wasn't aware it was deprecated. I've changed the base image to be Amazon Corretto now.

**Review by Jimexist — APPROVED** [review]

**Pull request merged**